### PR TITLE
Adds ability to set an account lockout policy

### PR DESCRIPTION
**Pull request by cherukumilli**
This is a fix for issue #2508 

# Account lockout policy:
Someone who attempts to use more than a few unsuccessful passwords while trying to log on to your system might be a malicious user who is attempting to determine an account password by trial and error.

Update parse-server to track logon attempts and respond to this type of potential attack by disabling the account for a preset period of time

This setting will have 2 parameters:
# Account lockout duration
The **Account lockout duration** policy setting determines the number of minutes that a locked-out account remains locked out before automatically becoming unlocked. The available range is from 1 through 99,999 minutes.

# Account lockout threshold 
The **Account lockout threshold** policy setting determines the number of failed sign-in attempts that will cause a user account to be locked. You can set a value from 1 through 999 failed sign-in attempts.

### CODE REVIEW

**Comment by cherukumilli** [comment]
@flovilmart 
If you get a few mins then can you please restart travis CI?

File: spec/AccountLockoutPolicy.spec.js [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+          reject(err);
+        }
+      });
+    }, waitTime);
+  });
+};
+
+describe("Account Lockout Policy: ", () => {
+
+  it_exclude_dbs(['postgres'])('account should not be locked even after failed login attempts if account lockout policy is not set', done => {
```
Comment (flovilmart) [inline spec/AccountLockoutPolicy.spec.js]: what's the problem with PG?

Comment (cherukumilli) [inline spec/AccountLockoutPolicy.spec.js]: I was under the impression that PG is not ready yet for testing and we are excluding test cases for PG.
Let me give the PG tests a shot and submit another pull request.

BTW: do you know if there is a workflow doc for contributing? i.e., a document that walks the contributor about the setup process, etc...

Comment (flovilmart) [inline spec/AccountLockoutPolicy.spec.js]: Not really, for PG, you need postgres > 9.4 with POSTGIS extensions loaded, and create the tables (see .travis.yml for the commands)

**Comment by facebook-github-bot-9** [comment]
@cherukumilli updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2601/files/173408b134e935e0d0e4b8474a884709dcfab4c5..c6af3647ee0fa565579a4f32a36c89e362c724a1)

**Comment by flovilmart** [comment]
Seems that there is an issue with PG on your PR. I'd still like to get it up and running for that DB.

you can debug the 3 failing tests 

- Account Lockout Policy:  lock account if failed login attempts are above threshold
- Account Lockout Policy:  lock account for accountPolicy.duration minutes if failed login attempts are above threshold
- Account Lockout Policy:  allow login for locked account after accountPolicy.duration minutes

by replacing the `it`  with a `fit` and also setting PARSE_SERVER_LOG_LEVEL=debug before running `npm test` like that:

``` 
PARSE_SERVER_LOG_LEVEL=debug PARSE_SERVER_TEST_DB=postgres npm test
```

**Comment by cherukumilli** [comment]
@flovilmart 
Many thanks for the feedback and setting the environment variables. I will push a PR tomorrow after fixing the code to work with PG.

**Is there a Contributor Workflow document that has the instructions like setting the environment variables?** I wonder if I should add the instructions you provided above to the existing contributions document (https://github.com/ParsePlatform/parse-server/blob/master/CONTRIBUTING.md). It will be useful for other contributors.

**Comment by flovilmart** [comment]
@cherukumilli you're 100% right! We should add that to this file. I'll add it

**Comment by facebook-github-bot-2** [comment]
@cherukumilli updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2601/files/c6af3647ee0fa565579a4f32a36c89e362c724a1..2d553e41d9c2aba7417342cbe8817b2ead9ca53c)

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 2]
```diff
@@ -651,8 +653,18 @@ export class PostgresStorageAdapter {
...
+        if (fieldName == '_account_lockout_expires_at') {
+          if (object[fieldName]) {
+            valuesArray.push(object[fieldName].iso);
+          } else {
+            valuesArray.push(null);
+          }
+        }
         if (fieldName == '_perishable_token') {
-          valuesArray.push(object[fieldName].iso);
+          valuesArray.push(object[fieldName].iso); // TODO: check with @flovilmart about why .iso is needed here
```
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: that is maybe an error, is the _perishable_token a date?

**Comment by cherukumilli** [comment]
@flovilmart 

Updating `it` to `fit` in the new `AccountLockoutPolicy.spec.js` file is causing only tests in this file to be executed by `npm test`. 

Also I don't see any of the other test cases using `fit`.

**Comment by facebook-github-bot-9** [comment]
@cherukumilli updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2601/files/2d553e41d9c2aba7417342cbe8817b2ead9ca53c..6dcad4c03268ce47ba245ffea0e0898f39a571df)

**Comment by facebook-github-bot-3** [comment]
@cherukumilli updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2601/files/6dcad4c03268ce47ba245ffea0e0898f39a571df..2a76ed1acb89460721abeb9a07b70a8629d0b531)

**Comment by nitrag** [comment]
Nice PR! Was just thinking about the need for this last night!

**Comment by facebook-github-bot-7** [comment]
@cherukumilli updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2601/files/2a76ed1acb89460721abeb9a07b70a8629d0b531..fbcc1ef63070926f2cc9e6dcc1ab2c99b4e52bda)

**Comment by flovilmart** [comment]
Looking good, but change fit for it (otherwise only fit tests run :P)

**Comment by facebook-github-bot-5** [comment]
@cherukumilli updated the pull request - [view changes](https://github.com/ParsePlatform/parse-server/pull/2601/files/fbcc1ef63070926f2cc9e6dcc1ab2c99b4e52bda..67b59429f13c21657efdb54c510872bd4aca6000)

**Comment by flovilmart** [comment]
Very neat feature as always @cherukumilli ! Thanks for your commitment to make it more secure!

**Comment by cherukumilli** [comment]
Thanks @flovilmart 

btw: I see warnings and errors like the following in my local setup for PG GeoPoint tests in file `ParseGeoPoint.spec.js`. Do I need to do anything special to my local setup to get the GeoPoint tests working?

```
[33mwarn[39m: Unable to ensure uniqueness for usernames:  code=107, message=schema class name does not revalidate
```

```
[33mwarn[39m: Unable to ensure uniqueness for user email addresses:  code=107, message=schema class name does not revalidate
```

```
F[33mwarn[39m: Unable to ensure uniqueness for usernames:  error: type "_SCHEMA" already exists
    at [object Object].Connection.parseE (/Users/diwakarcherukumilli/projects/parse/parse-server/node_modules/pg-promise/node_modules/pg/lib/connection.js:539:11)
    at [object Object].Connection.parseMessage (/Users/diwakarcherukumilli/projects/parse/parse-server/node_modules/pg-promise/node_modules/pg/lib/connection.js:366:17)
    at Socket.<anonymous> (/Users/diwakarcherukumilli/projects/parse/parse-server/node_modules/pg-promise/node_modules/pg/lib/connection.js:105:22)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:153:18)
    at Socket.Readable.push (_stream_readable.js:111:10)
    at TCP.onread (net.js:531:20)
```

Please see the attached file for error details after running command `npm test ParseGeoPoint.spec.js`

[ParseGeoPoint.Errors.Errors.zip](https://github.com/ParsePlatform/parse-server/files/453004/ParseGeoPoint.Errors.Errors.zip)

**Comment by flovilmart** [comment]
Seems ok on Travis. Did you enable the postgis extension?

**Comment by flovilmart** [comment]
Looking very nice!!  Thanks again!

**Comment by cherukumilli** [comment]
@flovilmart 

Is this going to be targeted for v2.3.0?
If so, is there a target date for 2.30?

**Comment by flovilmart** [comment]
We don't have a target date for the next release, as soon as all Pr are merged for 2.2.19 we'll release.

**Comment by ykdemandware** [comment]
Hi all. It seems like a locked out account although not having access to the login api still can access others classes/objects with a valid session token. Why don't we remove all the user's session token when the user is locked out?

**Comment by ederelk** [comment]
Hi all. It seems like a locked out account although not having access to the login api still can access others classes/objects with a valid session token. Why don't we remove all the user's session token when the user is locked out?

**Comment by ederelk** [comment]
Is there a way to actually lock the user out in cloud code?

**Comment by cherukumilli** [comment]
@ederelk 
IMO, allowing valid session tokens access to the api other than login api is acceptable behavior.

For example: 
User A logins from System X
User B (unauthorized user) tries to login from System Y **unsuccessfully** and ends up locking the account.

In the scenario above, IMO,
1. we should let User A continue to use their valid session token to use the application
2. lock the account so that User B is unable to login again for the lockout duration

Please let me know if you are seeing a security vulnerability that I am not seeing here.

**Comment by ahmedu007** [comment]
@cherukumilli  @facebook-github-bot @flavionegrao 
Is there any documentation explaining this please?
